lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,96 @@
+"""
+Middleware for Mobile APIs
+"""
+from datetime import datetime
+from django.conf import settings
+from django.http import HttpResponse
+import re
+from mobile_api.models import MobileAppVersionConfig
+from openedx.core.lib.mobile_utils import is_request_from_mobile_app
+
+
+class AppVersionUpgrade(object):
+    """
+    Middleware class to keep track of mobile application version being used
+    """
+    REGEX_USER_AGENT = {
+        'android': (r'Dalvik/[.0-9]+ \(Linux; U; Android [.0-9]+; '
+                    '(.*) Build/[0-9a-zA-Z]*\) (.*)/[0-9]+.[0-9]+.[0-9]+(.[0-9a-zA-Z]*)?'),
+        'ios': r'\([0-9]+.[0-9]+.[0-9]+(.[0-9a-zA-Z]*)?; OS Version [0-9.]+ \(Build [0-9a-zA-Z]*\)\)',
+    }
+
+    def version_tuple(self, version):
+        """ Converts string X.X.X to int tuple (X, X, X) """
+        return tuple(map(int, (version.split("."))))
+
+    def form_factor(self, user_agent):
+        """ extracts mobile platform and app version in use """
+        if re.search(self.REGEX_USER_AGENT['ios'], user_agent):
+            # version is found betweet first occurrence ( and ; i.e. (version;
+            user_app_version = user_agent[user_agent.find("(") + 1:user_agent.find(";")]
+            return "ios", user_app_version
+        elif re.search(self.REGEX_USER_AGENT['android'], user_agent):
+            # version is found as last entry in android user agent after "/"
+            user_app_version = user_agent[user_agent.rfind("/") + 1:]
+            return "android", user_app_version
+
+    def validate_version(self, user_app_version):
+        """ if user app version confronts to X.X.X.Y format return X.X.X else None"""
+        if re.match(settings.MOBILE_APP_VERSION_REGEX, user_app_version):
+            if user_app_version.count(".") == 3:
+                user_app_version = user_app_version[:user_app_version.rfind(".")]
+            return user_app_version
+
+    def process_request(self, request):
+        """
+        If request is from mobile native app and confront to X.X.X.Y version format (where X is a number and
+        Y is optional alphanumeric), then return Http 426 Upgrade Required if;
+        1. user app version < minimum supported version
+        2. user app version >= minimum supported version and user app version < next supported version and
+        request timestamp <= update required timespan for next supported version
+        """
+        if is_request_from_mobile_app(request):
+            user_agent = request.META.get('HTTP_USER_AGENT')
+            platform, user_app_version = self.form_factor(user_agent)
+
+            # if user app version does not match to X.X.X.Y (acceptable format) then let pass the request
+            # as it might belong to older apps
+            user_app_version = self.validate_version(user_app_version)
+            if user_app_version:
+                user_app_version_tuple = self.version_tuple(user_app_version)
```

Comment:
You can collapse the above 3 method calls into a method `parse_version_from_user_agent()` that returns `[platform, parsed_version]` if it can find a valid version or `None`.